enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,617 @@
+    // terminates encryption (that is, edge-terminated or reencrypt
+    // connections).  Headers cannot be set/append for TLS passthrough
+    // connections.
+    //
+    // +kubebuilder:validation:Optional
+    // +optional
+    Set *RouteSetHTTPHeader `json:"set,omitempty" protobuf:"bytes,9,opt,name=set"`
+
+    // delete defines HTTP headers that should be deleted.
+    // If this field is empty, no headers are deleted.
+    //
+    // Note that this option only applies to cleartext HTTP connections
+    // and to secure HTTP connections for which the ingress controller
+    // terminates encryption (that is, edge-terminated or reencrypt
+    // connections).  Headers cannot be set(replace)/append for TLS passthrough
+    // connections.
+    //
+    // +kubebuilder:validation:Optional
+    // +optional
+    Delete *RouteDeleteHTTPHeader `json:"delete,omitempty" protobuf:"bytes,10,opt,name=delete"`
+}
+```
+
+```go
+// RouteSetAction is used to understand as whether you want to perform Set related 
+// operation like Replace/Append Or a Delete operation
+// +kubebuilder:validation:Enum=Set;Delete
+type RouteSetHTTPHeaderAction string
+
+const (
+	Set    RouteSetHTTPHeaderAction = "Set"
+	Delete RouteSetHTTPHeaderAction = "Delete"
+)
+```
+The RouteSetHTTPHeader type has fields for specifying a header name, header value and action.
+```go
+type RouteSetHTTPHeader struct{
+    // Value specifies a header value.
+    // +kubebuilder:validation:Required
+    // +required
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=4096
+    Value string `json:"value"`
+
+    // Action specifies a header value.
+    // +kubebuilder:validation:Required
+    // +required
+    // +kubebuilder:validation:Enum=Append;Replace
+    Action RouteHTTPHeaderAction `json:"action"`
+}
+```
+
+```go
+// RouteHTTPHeaderAction is a policy for setting, replacing or appending HTTP headers.
+//
+// +kubebuilder:validation:Enum=Append;Replace
+type RouteHTTPHeaderAction string
+
+const (
+	// AppendHTTPHeaderPolicy appends the header, preserving any existing header.
+	AppendHTTPHeader RouteHTTPHeaderPolicy = "Append"
+	// ReplaceHTTPHeaderPolicy sets the header, removing any existing header.
+	ReplaceHTTPHeaderPolicy RouteHTTPHeaderPolicy = "Replace"
+)
+```
+
+The RouteDeleteHTTPHeader is an empty union member .
+```go
+type RouteDeleteHTTPHeader struct{
+}
+```
+
+### Implementation Details/Notes/Constraints
+A) Setting HTTP headers by cluster admin via IngressController's Spec.
+
+Implementing this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/cluster-ingress-operator
+* openshift/router
+
+The router configures HAProxy using a configuration template.  The template uses
+environment variables as input parameters.  The enhancement adds a new
+environment variables, `ROUTER_SET_HTTP_HEADERS`, which specify the headers to set in the
```

Comment:
I'm still confused about `ROUTER_SET_HEADERS` variable. Would it carry all the action types (`Append` and `Replace`)? How would you distinguish which headers need to be appended and which set?

Reply:
Yes.  We will append action along with `setheader name:setheader value: action`in the env var.

Reply:
As per [the design discussion](https://github.com/openshift/enhancements/pull/1296#discussion_r1065042615) we won't have append and replace actions, so no need to make a difference.

Reply:
No we will need the Action to be passed in env var of router deployment.

Reply:
We will pass Set or Delete

Reply:
I think I get your point, you mean we can just check if ROUTER_SET_HTTP_HEADERS is set , if yes then the Action is Set.